File: Code/C++/1626. Best Team With No Conflicts.cpp
#include<vector>
#include<algorithm>

using namespace std;


class Solution {
public:
    int bestTeamScore(vector<int>& scores, vector<int>& ages) {
        int max_age = *max_element(ages.begin(), ages.end());
        vector<int> dp(max_age+1, 0);
        vector<pair<int, int>> pairs;
        for (int i = 0; i < scores.size(); i++) {
            pairs.push_back({scores[i], ages[i]});
        }
        sort(pairs.begin(), pairs.end());
        for (auto &[score, age]: pairs){
            dp[age] = *max_element(dp.begin(), dp.begin() + age + 1) + score; 
        }
        return *max_element(dp.begin(), dp.end());
    }
};

File: Code/Python/1626. Best Team With No Conflicts.py
class Solution:
    def bestTeamScore(self, scores, ages) -> int:
        # dp[age]:= max scores up to age
        # O(nlogn + n*max_age)
        dp = [0]*(max(ages) + 1)
        for score, age in sorted(zip(scores, ages)):
            dp[age] = max(dp[:age+1]) + score

        return max(dp)


        # pairs = sorted(zip(ages, scores))
        # n = len(scores)
        # dp = [0]*n 
        # for i in range(n):
        #     dp[i] = pairs[i][1]
        #     for j in range(i):
        #         if pairs[j][1] <= pairs[i][1]:
        #             dp[i] = max(dp[i], dp[j] + pairs[i][1])
        # return max(dp)


              
            